backend/src/main/java/ai/verta/modeldb/batchProcess/CollaboratorResourceMigration.java

```diff
@@ -0,0 +1,236 @@
+            criteriaQuery.select(root).orderBy(criteriaBuilder.asc(root.get("id")));
+
+        TypedQuery<ProjectEntity> typedQuery = session.createQuery(selectQuery);
+
+        typedQuery.setFirstResult(lowerBound);
+        typedQuery.setMaxResults(pagesize);
+        List<ProjectEntity> projectEntities = typedQuery.getResultList();
+
+        if (projectEntities.size() > 0) {
+          Set<String> userIds = new HashSet<>();
+          for (ProjectEntity projectEntity : projectEntities) {
+            userIds.add(projectEntity.getOwner());
+          }
+          LOGGER.debug("Project userId list : " + userIds);
+
+          // Fetch the experiment owners userInfo
+          Map<String, UserInfo> userInfoMap =
+              authService.getUserInfoFromAuthServer(userIds, null, null);
+          for (ProjectEntity project : projectEntities) {
+            try {
+              WorkspaceDTO workspaceDTO =
+                  roleService.getWorkspaceDTOByWorkspaceId(
+                      userInfoMap.get(project.getOwner()),
+                      project.getWorkspace(),
+                      project.getWorkspace_type());
+              // if projectVisibility is not equals to ResourceVisibility.ORG_SCOPED_PUBLIC then
+              // ignore the CollaboratorType
+              createWorkspacePermissions(
+                  Optional.of(workspaceDTO.getWorkspaceName()),
+                  Optional.ofNullable(
+                      WorkspaceTypeEnum.WorkspaceType.forNumber(project.getWorkspace_type())),
+                  project.getId(),
+                  Optional.of(Long.parseLong(project.getOwner())),
+                  ModelDBServiceResourceTypes.PROJECT,
+                  CollaboratorTypeEnum.CollaboratorType.READ_WRITE,
+                  VisibilityEnum.Visibility.forNumber(project.getProject_visibility()));
+            } catch (Exception ex) {
+              LOGGER.warn("CollaboratorResourceMigration Exception: {}", ex.getMessage());
+            }
+          }
+        } else {
+          LOGGER.debug("Total projects count 0");
+        }
+        lowerBound += pagesize;
+      } catch (Exception ex) {
+        if (ModelDBUtils.needToRetry(ex)) {
+          migrateProjects();
+        } else {
+          throw ex;
+        }
+      }
+    }
+
+    LOGGER.debug("Projects migration finished");
+  }
+
+  private static Long getEntityCount(Class<?> klass) {
+    try (Session session = ModelDBHibernateUtil.getSessionFactory().openSession()) {
+      CriteriaBuilder criteriaBuilder = session.getCriteriaBuilder();
+      CriteriaQuery<Long> countQuery = criteriaBuilder.createQuery(Long.class);
+      countQuery.select(criteriaBuilder.count(countQuery.from(klass)));
+      return session.createQuery(countQuery).getSingleResult();
+    } catch (Exception ex) {
+      if (ModelDBUtils.needToRetry(ex)) {
+        return getEntityCount(klass);
+      } else {
+        throw ex;
+      }
+    }
+  }
+
+  private void createWorkspacePermissions(
```

Comment:
There's another method in another currently open PR with exactly the same name. cc @coryverta is this a copy?

Reply:
Yes it is a copy-paste method but there is one change which is https://github.com/VertaAI/modeldb/blob/fe66220573e9ebb12c2781b45257a825db853f0a/backend/src/main/java/ai/verta/modeldb/batchProcess/CollaboratorResourceMigration.java#L228 so when we are migrating if the project visibility has `Visibility.ORG_SCOPED_PUBLIC` then we are setting `ResourceVisibility.ORG_DEFAULT;` at UAC. but when we create project and getting visibility `Visibility.ORG_SCOPED_PUBLIC` then we are setting `ResourceVisibility.ORG_SCOPED_PUBLIC;` at tge UAC from RoleService. thats why define second method here.

Reply:
so createWorkspacePermissions looks the same as in roleService, but it uses another getResourceVisibility functionality? Can't we just pass the right visibility to roleService.createWorkspacePermissions?

Reply:
Ok, let me try if its possible.

Reply:
Yup, let me check if possible. i have duplicated it because createWorkspacePermissions called from multiple places and i don't want to touch those all for converting `Visivility.*` to `ResourceVisibility.*` logic multiple places.

Reply:
Let's not duplicate. I prefer to have this implemented well in the first pass than with considerable technical debt before it's even functional.